File: message-io/src/main/java/net/easyrpc/message/io/Transport.java
package net.easyrpc.message.io;

import org.jetbrains.annotations.NotNull;

import java.nio.channels.SocketChannel;
import java.util.concurrent.ConcurrentLinkedQueue;

public class Transport implements Comparable<Transport> {

    public SocketChannel channel;
    public Object attachment;
    public ConcurrentLinkedQueue<Event> taskQueue = new ConcurrentLinkedQueue<>();

    Transport(SocketChannel channel) {
        this.channel = channel;
    }

    public void send(String tag, Object object) {
        taskQueue.add(new Event(tag, object));
    }

    Transport attach(Object attachment) {
        this.attachment = attachment;
        return this;
    }

    @Override
    public int compareTo(@NotNull Transport o) {
        if (o.hashCode() == hashCode()) return 0;
        return (o.hashCode() > hashCode()) ? 1 : -1;
    }
}

File: message-io/src/main/kotlin/net/easyrpc/message/io/core/Transport.kt
package net.easyrpc.message.io.core

import net.easyrpc.message.io.api.ErrorHandler
import net.easyrpc.message.io.model.Event
import java.nio.channels.SocketChannel
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * @author chpengzh
 */
class Transport(var channel: SocketChannel,
                private var attachment: Any? = null,
                val error: ErrorHandler,
                val taskQueue: ConcurrentLinkedQueue<Event> = ConcurrentLinkedQueue()) :
        Comparable<Transport> {

    override fun compareTo(other: Transport): Int = when {
        other.hashCode() == hashCode() -> 0
        other.hashCode() < hashCode() -> -1
        else -> 1
    }

    fun send(tag: String, obj: Any) {
        taskQueue.add(Event(tag, obj))
    }

    fun attach(attachment: Any): Transport {
        this.attachment = attachment
        return this
    }

    fun attachment(): Any? {
        return attachment
    }

}